File: 2024-AoC/2-2.py
###
# Advent of Code 2024
# Day 2 - Part 2
#
# Answer: 717
#
# @author Kian Chua
#
##

def works(a):
    hasDec = False
    hasInc = False

    for i in range(1, len(a)):
        difference = abs(a[i] - a[i-1])
        if 1 > difference or difference > 3:
            return False
        elif difference == a[i] - a[i-1]:
            hasInc = True
        else:
            hasDec = True
    
    return hasInc ^ hasDec


def check():
    a = [int(x) for x in input().split()]

    if works(a):
        return True
    
    for i in range(len(a)):
        b = []
        for j in range(len(a)):
            if i != j:
                b.append(a[j])
        if works(b):
            return True

    return False


def main():
    n = 1000
    ans = 0

    for _ in range(n):
        ans += check()

    print(ans)


if __name__ == "__main__":
    main()

File: 2024-AoC/22.cpp
/**
 *  Advent of Code 2024
 *  Day 18 - Part 1 & Part 2
 * 
 *  Part 1 Answer: 324
 *  Part 2 Answer: 46,23
 * 
 *  @author Kian Chua
 * 
 */

#include <bits/stdc++.h>

#define ll long long
#define pb push_back
#define mp make_pair
#define f first
#define s second

using namespace std;

const int N = 2346;
const int T = 2000;

void solve() {

    ll part1 = 0;

    for (int _ = 0; _ < N; _++) {
        ll secret; cin >> secret;
        for (int i = 0; i < T; i++) { secret = generate(secret); }
        part1 += secret;
    }

    cout << part1 << endl;

    return;
}

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);

    int t = 1;
    while (t--) { solve(); }
}